index.js

```diff
@@ -1,79 +1,91 @@
 
-inherits(Dat, events.EventEmitter)
+    if(!this.opts.id) this.opts.id = randomBytes(32)
 
-/**
- * Returns a repo with the given key. Returns undefined
- * if no repository is found with that key.
- * @param  {string} key      The key of the repo.
- * @return {Repo|undefined}  The repo object with the corresponding key.
- */
-Dat.prototype.get = function (key) {
-  return this.repos.filter(function (repo) {
-    return key.toString('hex') === repo.key.toString('hex')
-  })[0]
-}
+    this.repos = []
+  }
 
-/**
- * Adds a new dat. Emits a 'repo' event when the undelying archive
- * instance is open.
- * @param {string}   key   The key to the dat.
- * @param {object}   opts  Options to use when building the dat.
- * @param {Function} cb    The callback with the repo object (optional).
- */
-Dat.prototype.add = function (key, opts, cb) {
-  var self = this
-  if (self.destroyed) throw new Error('client is destroyed')
-  if (typeof opts === 'function') return self.add(key, null, opts)
-  if (typeof key === 'function') return self.add(null, null, key)
-  if (!opts) opts = {}
-
-  var repo = new Repo(key, xtend(this.opts, opts))
-  self.repos.push(repo)
-
-  repo.once('ready', onready)
-  repo.once('close', onclose)
-
-  function onready () {
-    if (self.destroyed) return
-    if (typeof cb === 'function') cb(repo)
-    self.emit('repo', repo)
+  /**
+   * Returns a repo with the given url. Returns undefined
+   * if no repository is found with that url.
+   * @param  {url} url      The url of the repo.
+   * @return {Repo|undefined}  The repo object with the corresponding url.
+   */
+  get (url) {
+    const repo = this.repos.find((repo) => repo.url === url)
+    if (repo) return repo
+    return this.add(url)
   }
 
-  function onclose () {
-    repo.removeListener('ready', onready)
-    repo.removeListener('close', onclose)
+  /**
+   * Adds a new dat. Emits a 'repo' event when the undelying archive
+   * instance is open.
+   * @param {string}   url   The url to the dat.
+   * @param {object}   opts  Options to use when building the dat.
+   */
+  add (url, opts) {
+    if (this.destroyed) throw new Error('client is destroyed')
+    if (!opts) opts = {}
+
+    const repo = new Repo(url, xtend(this.opts, opts))
+    this.repos.push(repo)
+
+    repo.ready(() => {
+      this.emit('repo', repo)
+    })
+
+    return repo
   }
-}
 
-/**
- * Closes the dat, the swarm, and all underlying repo instances.
- */
-Dat.prototype.destroy =
-Dat.prototype.close = function () {
-  while (this.repos.length) {
-    var repo = this.repos.pop()
-    repo.close()
+  create (opts) {
+    return this.add(null, opts)
+  }
+
+  has (url) {
+    return !!this.repos.find((repo) => repo.url === url)
+  }
+
+  /**
+   * Closes the dat, the swarm, and all underlying repo instances.
+   */
+  close (cb) {
+    this.destroyed = true
+
+    if(cb) this.once('close', cb)
+
+    while (this.repos.length) {
+      const repo = this.repos.pop()
+      repo.close()
```

Comment:
This could be done in a different PR because you're just keeping feature parity with the old code, but perhaps it might be best to wait for all the repos to close before we call the cb?

Reply:
could use something like run-parallel for this

Reply:
Good point! This could be a prime suspect for race conditions. :D